GeneratingPolicy with generateExisting panics background controller when matchConstraints targets a resource the SA cannot list

# [Bug] GeneratingPolicy with `generateExisting` panics background controller when `matchConstraints` targets a resource the SA cannot list

## Kyverno Version

1.17.0 , 1.18.0 and older versions

## Kubernetes Version

1.34.x

## Kubernetes Platform

EKS

## Kyverno Rule Type

Generate

---

## Description

When a `GeneratingPolicy` (`policies.kyverno.io/v1`) has `spec.evaluation.generateExisting.enabled: true` and `matchConstraints.resourceRules` targets `customresourcedefinitions`, the background controller calls `ListResource` on CRDs cluster-wide. However, the `kyverno-background-controller` ServiceAccount only has `get` on CRDs — not `list` or `watch` — so the API server returns `403 Forbidden`.

The upstream bug is in `pkg/policy/gpol.go:103–107`: the error branch logs but is missing a `continue` statement, so execution falls through and immediately dereferences the nil result, causing:

```
runtime error: invalid memory address or nil pointer dereference
```

Both background controller replicas enter `CrashLoopBackOff`, which causes significant disruption as the background controller is responsible for processing all generate rules.

**Notes on scope:**

- The bug only fires with `generateExisting: true`. Without it, applying the policy doesn't crash the controller.
- The bug fires even when no matching CRD exists in the cluster. `ListResource` is called before any `matchConditions` are evaluated, so the cluster-wide list itself is what fails.
- A `403 Forbidden` from the API server is something controllers encounter regularly (RBAC narrowed, CRD removed mid-reconcile, etc.). It should not cause a panic.
- The default Kyverno chart grants only `get` on CRDs to the background controller, so a `GeneratingPolicy` triggering on CRDs crashes the controller out of the box.

> Related: #15722 — a separate `generateExisting` issue (UR race condition causing generated resources to be deleted immediately after creation). Same surface area, different code path; the fixes are independent.

---

## Steps to Reproduce

Install Kyverno via the official Helm chart 3.7.x (Kyverno 1.18.0) with default RBAC, then apply:

```yaml
apiVersion: policies.kyverno.io/v1
kind: GeneratingPolicy
metadata:
  name: repro-gpol-panic
spec:
  evaluation:
    generateExisting:
      enabled: true
  matchConstraints:
    resourceRules:
      - apiGroups: ["apiextensions.k8s.io"]
        apiVersions: ["v1"]
        operations: ["CREATE", "UPDATE"]
        resources: ["customresourcedefinitions"]
  matchConditions:
    - name: never-matches
      expression: "object.metadata.name == 'crd-that-does-not-exist'"
  generate:
    - expression: |-
        generator.Apply("", [
          {
            "apiVersion": dyn("v1"),
            "kind": dyn("ConfigMap"),
            "metadata": dyn({
              "name": dyn("would-be-generated"),
              "namespace": dyn("default")
            })
          }
        ])
```

The cluster does not need to have any matching CRD — the `LIST` fails before any matching happens. `matchConditions` is set to a name that doesn't exist so the policy stays harmless if the bug is fixed.

Within ~60 seconds the background controller starts crashlooping. Confirm via:

```bash
kubectl logs -n kyverno \
  -l app.kubernetes.io/component=background-controller \
  --previous --tail=200
```

The stack trace points at `pkg/policy/gpol.go:107`.

---

## Expected Behavior

`getGpolTriggers` should skip the failing GVR and continue with the rest of the trigger rules. Optionally, emit a `PolicyError` event so the missing RBAC is visible to operators.

---

## Suggested Fix

Add `continue` after the error log:

```go
resources, err := pc.client.ListResource(context.TODO(), groupVersion.String(), gvk.Kind, "", objectSelector)
if err != nil {
    pc.log.Error(err, "failed to list resources", "groupVersion", groupVersion, "kind", gvk.Kind)
+   continue
}
for i, res := range resources.Items {
```

The same pattern is already used for the `restMapper.KindFor` error branch a few lines up.

Additionally, the default Helm chart RBAC for `kyverno-background-controller` should be updated to include `list` and `watch` on `customresourcedefinitions`, since triggering on CRDs is a common use case.

---

## Kyverno Logs

```
ERR github.com/kyverno/kyverno/pkg/policy/gpol.go:105 > failed to list resources
  error="customresourcedefinitions.apiextensions.k8s.io is forbidden:
  User \"system:serviceaccount:kyverno:kyverno-background-controller\"
  cannot list resource \"customresourcedefinitions\" in API group
  \"apiextensions.k8s.io\" at the cluster scope"
  groupVersion=apiextensions.k8s.io/v1 kind=CustomResourceDefinition
  logger=PolicyController

panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x3dcaa13]

goroutine 396 [running]:
...
github.com/kyverno/kyverno/pkg/policy.(*policyController).getGpolTriggers(0xc000cd5800, 0xc000583220)
        github.com/kyverno/kyverno/pkg/policy/gpol.go:107 +0x513
github.com/kyverno/kyverno/pkg/policy.(*policyController).handleGenerateExisting(0xc000cd5800, 0xc00004bd40)
        github.com/kyverno/kyverno/pkg/policy/gpol.go:63 +0x10c
github.com/kyverno/kyverno/pkg/policy.(*policyController).syncPolicy(0xc000cd5800, {0xc008cb4990, 0x24})
        github.com/kyverno/kyverno/pkg/policy/policy_controller.go:441 +0x865
```

---

## Workaround

Grant `list` and `watch` on `customresourcedefinitions` to the `kyverno-background-controller` ServiceAccount via an additive `ClusterRole`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: kyverno:background-controller:additional-permissions
rules:
  - apiGroups: ["apiextensions.k8s.io"]
    resources:
      - customresourcedefinitions
    verbs:
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kyverno:background-controller:additional-permissions
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kyverno:background-controller:additional-permissions
subjects:
  - kind: ServiceAccount
    name: kyverno-background-controller
    namespace: kyverno
```

This unblocks the CRD trigger case but must be repeated per resource type for any other non-listable resource. The two-line code fix above is the real resolution.

---

## Slack Discussion

_No response_

---

## Troubleshooting

- [x] I have read and followed the documentation AND the [[troubleshooting guide](https://kyverno.io/docs/troubleshooting/)](https://kyverno.io/docs/troubleshooting/).
- [x] I have searched other issues in this repository and mine is not recorded.

--- a/pkg/policy/gpol.go
+++ b/pkg/policy/gpol.go
@@ -103,6 +103,7 @@ func (pc *policyController) getGpolTriggers(match *admissionregistrationv1.Match
 					resources, err := pc.client.ListResource(context.TODO(), groupVersion.String(), gvk.Kind, "", objectSelector)
 					if err != nil {
 						pc.log.Error(err, "failed to list resources", "groupVersion", groupVersion, "kind", gvk.Kind)
+						continue
 					}
 					for i, res := range resources.Items {
 						if !pc.triggerMatches(res, gvr, rule.ResourceNames, match.ExcludeResourceRules, nsSelector) {

